src/sentry/ratelimits/cardinality.py

```diff
@@ -0,0 +1,139 @@
+"""
+
+
+import time
+from dataclasses import dataclass
+from typing import Any, Collection, Iterator, Optional, Protocol, Sequence, Tuple, TypeVar
+
+from sentry.utils.services import Service
+
+Hash = str
+Timestamp = int
+
+
+@dataclass(frozen=True)
+class Quota:
+    # The number of seconds to apply the limit to.
+    window_seconds: int
+
+    # A number between 1 and `window_seconds`. Since `window_seconds` is a
+    # sliding window, configure what the granularity of that window is.
+    #
+    # If this is equal to `window_seconds`, the quota resets to 0 every
+    # `window_seconds`.  If this is a very small number, the window slides
+    # "more smoothly" at the expense of having much more redis keys.
+    #
+    # The number of redis keys required to enforce a quota is `window_seconds /
+    # granularity_seconds`.
+    granularity_seconds: int
+
+    #: How many units are allowed within the given window.
+    limit: int
+
+    def __post__init__(self) -> None:
+        assert self.window_seconds % self.granularity_seconds == 0
+
+
+@dataclass(frozen=True)
+class RequestedQuota:
+    # A string that all redis state is prefixed with. For example
+    # `sentry-string-indexer` where 123 is an organization id.
+    prefix: str
+
+    # A unit is an abstract term for the object type we want to limit the
+    # cardinality of.
+    #
+    # For example, if you want to limit the cardinality of timeseries in a
+    # metrics service, this would be a set of hashes composed from `(org_id,
+    # metric_name, tags)`.
+    #
+    # ...though you can probably omit org_id if it is already in the prefix.
+    unit_hashes: Collection[Hash]
+
+    # Which quotas to check against. The number of not-yet-seen hashes must
+    # "fit" into all quotas.
+    quotas: Sequence[Quota]
+
+
+@dataclass(frozen=True)
+class GrantedQuota:
+    # The prefix from RequestedQuota
+    prefix: str
+
+    # The subset of hashes that passed through.
+    granted_unit_hashes: Collection[Hash]
+
+    # If len(granted_unit_hashes) < len(RequestedQuota.unit_hashes), this
+    # contains the quotas that were reached.
+    reached_quotas: Sequence[Quota]
+
+
+class CardinalityLimiter(Service):
+    def __init__(self, **options: Any) -> None:
+        pass
+
+    def check_within_quotas(
+        self, requests: Sequence[RequestedQuota], timestamp: Optional[Timestamp] = None
+    ) -> Tuple[Timestamp, Sequence[GrantedQuota]]:
+        """
+        Given a set of quotas requests and limits, compute how much quota could
+        be consumed.
+
+        :param requests: The requests to return "grants" for.
+        :param timestamp: The timestamp of the incoming request. Defaults to
+            the current timestamp.
+
+            Providing a too old timestamp here _can_ effectively disable rate
+            limits, as the older request counts may no longer be stored.
+            However, consistently providing old timestamps here will work
+            correctly.
```

Comment:
What about just not allowing the call site to provide the timestamp ?

Reply:
I copied this design decision + doccomment from the sliding windows rate limiter. There, we are doing checks and usage of rate limits in two separate steps. In order to have the same timestamp between those two steps, I need to return the timestamp from the check-step, and require it for the second step.

With the sliding-windows rate limiter we hide the timestamp and other state in a context manager, but that's a layer around this one.

Reply:
RIght, makes sense.